lib/ansible/playbook/base.py

```diff
@@ -304,6 +306,215 @@ def validate(self, all_vars=None):
+                    if defaults_entry.startswith('ansible.legacy.'):
+                        action_names.append(defaults_entry.replace('ansible.legacy.', 'ansible.builtin.'))
+
+                    # Replace the module_defaults action entry with the canonical name,
+                    # so regardless of how the action is called, the defaults will apply
+                    for action_name in action_names:
+                        resolved_action = self._resolve_action(action_name)
+                        if resolved_action:
+                            validated_defaults_dict[resolved_action] = defaults
+
+            validated_module_defaults.append(validated_defaults_dict)
+
+        return validated_module_defaults
+
+    @property
+    def play(self):
+        if hasattr(self, '_play'):
+            play = self._play
+        elif hasattr(self, '_parent') and hasattr(self._parent, '_play'):
+            play = self._parent._play
+        else:
+            play = self
+
+        if play.__class__.__name__ != 'Play':
+            # Should never happen, but handle gracefully by returning None, just in case
+            return None
+
+        return play
+
+    def _resolve_group(self, group, collection_name, mandatory=True):
+        # The group should be part of the current collection
+        if not group.startswith(collection_name + '.'):
+            fq_group_name = collection_name + '.' + group
+        else:
+            fq_group_name = group
+
+        # Check if the group has already been resolved and cached
+        if fq_group_name in self.play._group_actions:
+            return fq_group_name, self.play._group_actions[fq_group_name]
+
+        try:
+            action_groups = _get_collection_metadata(collection_name).get('action_groups', {})
+        except ValueError:
+            if not mandatory:
+                display.vvvvv("Error loading module_defaults: could not resolve the module_defaults group %s" % fq_group_name)
+                return fq_group_name, []
+
+            raise AnsibleParserError("Error loading module_defaults: could not resolve the module_defaults group %s" % fq_group_name)
+
+        # The collection may or may not use the fully qualified name
+        # Don't fail if the group doesn't exist in the collection
+        short_name = fq_group_name.split(collection_name + '.')[-1]
+        action_group = action_groups.get(
+            fq_group_name,
+            action_groups.get(short_name)
+        )
+        if action_group is None:
+            if not mandatory:
+                display.vvvvv("Error loading module_defaults: could not resolve the module_defaults group %s" % fq_group_name)
+                return fq_group_name, []
+            raise AnsibleParserError("Error loading module_defaults: could not resolve the module_defaults group %s" % fq_group_name)
+
+        resolved_actions = []
+        include_groups = []
+
+        found_group_metadata = False
```

Comment:
Feels like this validation stuff should be in a utility method (if not in a separate file); the validation code seems to dwarf the actual working code in the loop and obscures the "important" parts, and I'm guessing the McCabe number for this file is not in a happy place right now ;)

Reply:
I was not sure where to stick new general collection utilities. What do you think about lib/ansible/utils/collection_loader/_collection_meta.py?

Reply:
Moved it back into the base class, just as a standalone function